applications/nrf_desktop/doc/hid_reportq.rst

```diff
@@ -0,0 +1,81 @@
+.. _nrf_desktop_hid_reportq:
+
+HID report queue utility
+########################
+
+.. contents::
+   :local:
+   :depth: 2
+
+A HID subscriber (for example, a USB HID class instance handled by :ref:`nrf_desktop_usb_state`) limits the maximum number of HID input reports that can be handled simultaneously.
+Enqueuing reports locally at the source may be necessary to prevent HID report drops.
+The HID report queue utility can be used by an application module to simplify enqueuing HID input reports received from connected HID peripherals before providing them to HID subscriber.
+
+Configuration
+*************
+
+Make sure that heap size (:kconfig:option:`CONFIG_HEAP_MEM_POOL_SIZE`) is large enough.
+Data structures used to enqueue HID reports are dynamically allocated using the :c:func:`k_malloc` function.
+When no longer needed, the structures are freed using the :c:func:`k_free` function.
+
+Use the :ref:`CONFIG_DESKTOP_HID_REPORTQ <config_desktop_app_options>` Kconfig option to enable the utility.
+You can use the utility only on HID dongles (:ref:`CONFIG_DESKTOP_ROLE_HID_DONGLE <config_desktop_app_options>`).
+
+You can also configure the following properties:
+
+* Maximum number of enqueued HID reports (:ref:`CONFIG_DESKTOP_HID_REPORTQ_MAX_ENQUEUED_REPORTS <config_desktop_app_options>`)
+* Number of supported HID report queues (:ref:`CONFIG_DESKTOP_HID_REPORTQ_QUEUE_COUNT <config_desktop_app_options>`)
+
+See Kconfig help for more details.
+
+Using HID report queue
+**********************
+
+The utility acts as a middleware between a connected HID peripheral and a HID subscriber.
+The following sections describe how to integrate HID report queue APIs in an application module.
+
+The :ref:`nrf_desktop_hid_forward` is an example of an application module that uses the HID report queue utility.
+Refer to the module's implementation (:file:`src/modules/hid_forward.c`) for an example of HID report queue integration.
+
+Allocation
+==========
+
+A HID report queue instance must be allocated with the :c:func:`hid_reportq_alloc` function before it is used.
+A separate HID report queue object needs to be allocated for each HID report subscriber.
+During allocation, the caller specifies the HID subscriber identifier.
+The :c:func:`hid_reportq_get_sub_id` function can be used to access the identifier later.
+After the HID report queue object is no longer used, it should be freed using the :c:func:`hid_reportq_free` function.
+
+HID subscriptions
+=================
+
+A HID report queue instance tracks HID input report subscriptions enabled by a HID subscriber.
+You can use the :c:func:`hid_reportq_subscribe` and :c:func:`hid_reportq_unsubscribe` functions to update the state of enabled subscriptions.
```

Comment:
Shouldn't it be the other way that:
 `the subscribers module can subscribe to reportq using ....`

Reply:
Or the hid_forward module which is a middleware module???

Reply:
HID forward integrates `hid_reportq`. HID forward uses the `hid_reportq_subscribe` and `hid_reportq_unsubscribe` to track subscriptions of a handled HID transport/subscriber (USB state module). The functions are not used directly by the USB state module.

Reply:
HID report queue acts as a middleware between HID forward and USB state